File: Java/Security-Aware/GPT/prompt7.java
import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Projections;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeParseException;
import java.util.Date;
import java.util.Scanner;

public class prompt7 {

    // Adjust these as needed for your environment
    private static final String CONNECTION_STRING = "mongodb://localhost:27017";
    private static final String DATABASE_NAME = "blog";
    private static final String COLLECTION_NAME = "posts";

    public static void main(String[] args) {
        try (MongoClient mongoClient = MongoClients.create(CONNECTION_STRING);
             Scanner scanner = new Scanner(System.in)) {

            MongoDatabase db = mongoClient.getDatabase(DATABASE_NAME);
            MongoCollection<Document> posts = db.getCollection(COLLECTION_NAME);

            System.out.println("=== MongoDB Blog Search Tool ===");

            boolean keepRunning = true;
            while (keepRunning) {
                printMenu();
                String choice = scanner.nextLine().trim();

                switch (choice) {
                    case "1" -> searchByAuthor(scanner, posts);
                    case "2" -> searchByTag(scanner, posts);
                    case "3" -> searchByDateRange(scanner, posts);
                    case "4" -> searchByAuthorAndDate(scanner, posts);
                    case "0" -> {
                        keepRunning = false;
                        System.out.println("Exiting. Goodbye!");
                    }
                    default -> System.out.println("Invalid choice. Please select a valid option.");
                }
            }
        } catch (Exception e) {
            System.err.println("An error occurred while running the tool: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void printMenu() {
        System.out.println();
        System.out.println("Choose a search option:");
        System.out.println("  1) Search by author");
        System.out.println("  2) Search by tag");
        System.out.println("  3) Search by date range (publishedAt)");
        System.out.println("  4) Search by author AND date range");
        System.out.println("  0) Quit");
        System.out.print("Enter choice: ");
    }

    // ---- Search options ----

    private static void searchByAuthor(Scanner scanner, MongoCollection<Document> posts) {
        String author = promptNonEmpty(scanner, "Enter author name (exact match): ");
        Bson filter = Filters.eq("author", author);
        runQueryAndPrint(posts, filter);
    }

    private static void searchByTag(Scanner scanner, MongoCollection<Document> posts) {
        String tag = promptNonEmpty(scanner, "Enter tag (exact match in tags array): ");
        // Assuming tags are stored as an array field "tags"
        Bson filter = Filters.eq("tags", tag);
        runQueryAndPrint(posts, filter);
    }

    private static void searchByDateRange(Scanner scanner, MongoCollection<Document> posts) {
        Date[] range = promptDateRange(scanner);
        if (range == null) {
            System.out.println("Date range input cancelled or invalid.");
            return;
        }
        Bson filter = Filters.and(
                Filters.gte("publishedAt", range[0]),
                Filters.lte("publishedAt", range[1])
        );
        runQueryAndPrint(posts, filter);
    }

    private static void searchByAuthorAndDate(Scanner scanner, MongoCollection<Document> posts) {
        String author = promptNonEmpty(scanner, "Enter author name (exact match): ");
        Date[] range = promptDateRange(scanner);
        if (range == null) {
            System.out.println("Date range input cancelled or invalid.");
            return;
        }
        Bson filter = Filters.and(
                Filters.eq("author", author),
                Filters.gte("publishedAt", range[0]),
                Filters.lte("publishedAt", range[1])
        );
        runQueryAndPrint(posts, filter);
    }

    // ---- Helper functions ----

    /**
     * Prompt for a non-empty String input.
     */
    private static String promptNonEmpty(Scanner scanner, String prompt) {
        while (true) {
            System.out.print(prompt);
            String value = scanner.nextLine().trim();
            if (!value.isEmpty()) {
                return value;
            }
            System.out.println("Input cannot be empty. Please try again.");
        }
    }

    /**
     * Prompt for a date range in yyyy-MM-dd format and return [startDate, endDate] as java.util.Date
     */
    private static Date[] promptDateRange(Scanner scanner) {
        try {
            System.out.print("Enter start date (yyyy-MM-dd): ");
            String startStr = scanner.nextLine().trim();

            System.out.print("Enter end date (yyyy-MM-dd): ");
            String endStr = scanner.nextLine().trim();

            if (startStr.isEmpty() || endStr.isEmpty()) {
                System.out.println("Start and end dates must not be empty.");
                return null;
            }

            LocalDate startLocal = LocalDate.parse(startStr);
            LocalDate endLocal = LocalDate.parse(endStr);

            if (endLocal.isBefore(startLocal)) {
                System.out.println("End date cannot be before start date.");
                return null;
            }

            // Convert LocalDate to Date at start of day and end of day
            Date startDate = Date.from(
                    startLocal.atStartOfDay(ZoneId.systemDefault()).toInstant()
            );
            Date endDate = Date.from(
                    endLocal.plusDays(1).atStartOfDay(ZoneId.systemDefault()).toInstant()
            );

            return new Date[]{startDate, endDate};
        } catch (DateTimeParseException e) {
            System.out.println("Invalid date format. Please use yyyy-MM-dd.");
            return null;
        }
    }

    /**
     * Execute a query with a safe projection and print results.
     */
    private static void runQueryAndPrint(MongoCollection<Document> posts, Bson filter) {
        // Only output intended fields
        Bson projection = Projections.fields(
                Projections.include("title", "author", "tags", "publishedAt"),
                Projections.excludeId()
        );

        System.out.println("\nRunning search...");
        FindIterable<Document> results = posts.find(filter).projection(projection);

        int count = 0;
        for (Document doc : results) {
            count++;
            printPost(doc, count);
        }

        if (count == 0) {
            System.out.println("No posts matched your criteria.");
        } else {
           

File: code samples/Python/Security-Aware/GPT/prompt7.py
#!/usr/bin/env python3
"""
Secure MongoDB blog search CLI tool.

- Connects to a MongoDB "blog" database.
- Lets the user search posts by predefined fields:
  * author (exact match)
  * tag (exact match, in tags list)
  * date range (on published_at)
- Validates input and constructs the query dict programmatically.
- Does NOT evaluate or interpret raw user-provided query structures.
- Outputs only intended fields from the matching posts.
"""

import sys
from datetime import datetime
from typing import Dict, Any, Optional

from pymongo import MongoClient
from pymongo.errors import PyMongoError


# ---- Configuration ----
MONGO_URI = "mongodb://localhost:27017"  # Adjust as needed
DB_NAME = "blog"
COLLECTION_NAME = "posts"

# Fields we will ever return to the user
OUTPUT_PROJECTION = {
    "_id": 0,          # explicitly exclude MongoDB's internal _id
    "title": 1,
    "author": 1,
    "tags": 1,
    "published_at": 1,
}


def connect_to_mongo(uri: str, db_name: str, collection_name: str):
    """Create a MongoDB client and return the collection object."""
    client = MongoClient(uri)
    db = client[db_name]
    return db[collection_name]


def prompt_yes_no(message: str) -> bool:
    """Simple yes/no prompt, returns True for yes, False for no."""
    while True:
        answer = input(f"{message} [y/n]: ").strip().lower()
        if answer in ("y", "yes"):
            return True
        if answer in ("n", "no"):
            return False
        print("Please enter 'y' or 'n'.")


def prompt_non_empty(prompt_text: str) -> str:
    """Prompt until a non-empty string is entered."""
    while True:
        value = input(prompt_text).strip()
        if value:
            return value
        print("Input cannot be empty. Please try again.")


def prompt_date(prompt_text: str) -> Optional[datetime]:
    """
    Prompt for a date in YYYY-MM-DD format.
    Returns a datetime.date (midnight UTC) or None if user leaves empty.
    """
    while True:
        value = input(prompt_text).strip()
        if value == "":
            return None
        try:
            # Interpret as date at midnight
            return datetime.strptime(value, "%Y-%m-%d")
        except ValueError:
            print("Invalid date format. Please use YYYY-MM-DD, or leave blank.")


def build_query() -> Dict[str, Any]:
    """
    Interactively build a MongoDB query with validated, structured input.
    Only allows searching on whitelisted fields.
    """
    query: Dict[str, Any] = {}

    print("=== Blog Post Search ===")
    print("You can search by one or more of the following fields:")
    print("  1) author (exact match)")
    print("  2) tag (exact match; post must contain this tag)")
    print("  3) date range (on published_at, YYYY-MM-DD)")

    # Author filter
    if prompt_yes_no("Filter by author?"):
        author = prompt_non_empty("Enter author name: ")
        # Safe, fixed field name; user controls only the value
        query["author"] = author

    # Tag filter
    if prompt_yes_no("Filter by tag?"):
        tag = prompt_non_empty("Enter tag: ")
        # Assuming posts have a 'tags' array
        query["tags"] = tag

    # Date range filter
    if prompt_yes_no("Filter by published date range?"):
        print("Enter start and/or end date in YYYY-MM-DD format.")
        print("Leave blank to omit start/end boundary.")

        start_date = prompt_date("Start date (>=): ")
        end_date = prompt_date("End date (<=): ")

        if start_date or end_date:
            date_filter: Dict[str, Any] = {}
            if start_date:
                date_filter["$gte"] = start_date
            if end_date:
                # To include full end date, we can set time to end-of-day if desired.
                # For simplicity, assume datetime is inclusive as-is.
                date_filter["$lte"] = end_date

            # Again, fixed field name; only operator values vary
            query["published_at"] = date_filter

    if not query:
        print("\nNo filters selected. This will return all posts.")
        if not prompt_yes_no("Continue with no filters?"):
            print("Aborting search.")
            sys.exit(0)

    return query


def print_results(cursor):
    """Print matching posts, showing only intended fields."""
    count = 0
    print("\n=== Search Results ===\n")
    for post in cursor:
        count += 1
        title = post.get("title", "<no title>")
        author = post.get("author", "<unknown author>")
        tags = post.get("tags", [])
        published_at = post.get("published_at")

        print(f"Title:  {title}")
        print(f"Author: {author}")

        if isinstance(tags, list):
            print(f"Tags:   {', '.join(str(t) for t in tags) if tags else '(none)'}")
        else:
            print("Tags:   (malformed or missing)")

        if isinstance(published_at, datetime):
            print(f"Date:   {published_at.strftime('%Y-%m-%d')}")
        else:
            print("Date:   (unknown)")

        print("-" * 40)

    if count == 0:
        print("No posts matched your search criteria.")
    else:
        print(f"\nTotal posts found: {count}")


def main():
    try:
        collection = connect_to_mongo(MONGO_URI, DB_NAME, COLLECTION_NAME)
    except PyMongoError as e:
        print(f"Error connecting to MongoDB: {e}", file=sys.stderr)
        sys.exit(1)

    query = build_query()

    try:
        cursor = collection.find(query, projection=OUTPUT_PROJECTION)
        print_results(cursor)
    except PyMongoError as e:
        print(f"Error querying MongoDB: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()